### DISCUSSION TITLE
Issue with distributing hello_world_distributed on several nodes

### DISCUSSION THREAD

**Discussion by dkeyGit** [STEllAR-GROUP/hpx#6307] [General]
Hello,

my university project group and me have some issues in relation to run the “hello_world_distributed” program. We want to run this program with SLURM on two nodes and the partition is called “qdr”. We start the running process with the following command: 

srun -p qdr -N 2 hello_world_distributed

Unfortunately, we get the following error message per node: 

Could not find usable bootstrap parcelport. 
Preconfigured bootstrap parcelport: 'tcp' 
List of available parcelports: 
 mpi, priority: -100, can bootstrap: 1

With the following command we get the debug information:
srun -p qdr -N 2 hello_world_distributed --hpx:list-parcel-ports --hpx:debug-clp

The output is here:
[output.txt](https://github.com/STEllAR-GROUP/hpx/files/12138934/output.txt)

We already searched for possible solutions and found an issue entry in https://github.com/STEllAR-GROUP/hpx/issues/4891. There we found the tip to expand the srun-command with the following options:

srun -p qdr -N 4 hello_world_distributed --hpx:ini=hpx.parcel.mpi.enable!=1 --hpx:ini=hpx.parcel.bootstrap!=mpi

Unfortunately, now we get the error message:

terminate called without an active exception 
Could not find usable bootstrap parcelport. 
Preconfigured bootstrap parcelport: 'mpi' 
List of available parcelports: 
 mpi, priority: -100, can bootstrap: 1

Here are our CmakeLists.txt and load-env.sh:
[CMakeLists.txt](https://github.com/STEllAR-GROUP/hpx/files/12138915/CMakeLists.txt)
[load-env.txt](https://github.com/STEllAR-GROUP/hpx/files/12138920/load-env.txt)

Do you have any ideas how to fix the errors?

**Comment by hkaiser** [comment]
You will need to launch your application using mpiexec/mpirun If you want to use the mpi parcelport. A good example on how to do this can be found here: https://github.com/STEllAR-GROUP/hpx/wiki/Running-HPX-on-Rostam#running-hpx-applications-on-compute-nodes. So you should probably use something like:
```
salloc -p qdr -N 2 mpirun ./bin/hello_world_distriuted
```

**Reply by dkeyGit** [reply]
Thank you, with this command the described error is fixed! But now we get the problem, that the program is only distributed among the threads of one locality (locality 0), other nodes are not involved. 

With: 
salloc -p qdr -N 4 -n 8 -c 8 mpirun hello_world_distributed 

we get the following output:
[output.txt](https://github.com/STEllAR-GROUP/hpx/files/12139121/output.txt)

Do you have any suggestions why it is not distributed among more localities?

**Reply by pauleonix** [reply]
@hkaiser How does this fit together with the fact that using `mpirun` is [discouraged](https://slurm.schedmd.com/mpi_guide.html#open_mpi) when using SLURM-aware OpenMPI? When installing SLURM-aware OpenMPI with spack, these wrappers are even [removed by default](https://github.com/spack/spack/pull/10340#issuecomment-454355612).

**Reply by hkaiser** [reply]
That's what I meant when I said that "all depends on how your slurm/MPI integration was built and configured". You don't need mpirun/mpiexec if your slurm installation initializes the required MPI environment variables.

Could you give me whatever `srun -p qdr -N 2 printenv` produces, please?

**Reply by ConstantinNicolai** [reply]
Hi, I'm also part of the project group, I attached a file with the output of this command on the cluster-headnode
[env.txt](https://github.com/STEllAR-GROUP/hpx/files/12172732/env.txt)

**Reply by hkaiser** [reply]
Yep, that's what I thought. There are no MPI related environment variables set on your system. That prevents MPI from initializing properly. I'm at a loss what to do.

Can you run a simple MPI program (unrelated to HPX) by just launching it without mpirun/mpiexec?

**Reply by pauleonix** [reply]
Yes, I can run e.g. `srun -p qdr -N 2 ./mpi_hello_world` with no problem and get greeted from 2 processes on two different qdr nodes. My understanding is that OpenMPI communicates with SLURM through PMI (our SLURM installation was built without PMIx), but maybe it is also just parses SLURM environment variables?

**Reply by hkaiser** [reply]
I understand. In this case, HPX can not be used to run with the MPI parcelport right now as it expects to see the MPI environment variables. I'll see what I can do to rectify this situation. Let's try one more thing: can you give me the output of `salloc -p qdr -N 2 mpirun -n 2 printenv`, please?

Now, that I think of it more. HPX relies on the usual MPI libraries to initialize MPI itself. So however the system is configured, the MPI libraries should pick it up correctly. Are you sure you built HPX against the same MPI version as your SLURM environment assumes (check with `ldd`)?

**Reply by pauleonix** [reply]
For me this just hangs without ever getting past the `mpirun`, probably some sort of ssh/kerberos authentification issue. But the students seem to have found a way to get it to work with hpx (i.e. `salloc` with `mpirun`). They have not told me the details yet though.

**Reply by hkaiser** [reply]
I'd love to hear how you resolved the issue.

**Reply by pauleonix** [reply]
"False negative", the processes were running on the headnode.

**Reply by dkeyGit** [reply]
Unfortunately it still does not work. With "salloc -p qdr -N 2 mpirun -n 2 hello_world_distributed --hpx:threads 2", we get the following output:
[hello_world_distributed_sample.txt](https://github.com/STEllAR-GROUP/hpx/files/12186208/hello_world_distributed_sample.txt)

We thought, that the program is distributed among 2 nodes (because there is "hello world" from 2 localities) on the qdr-partition, but after an analysis it turned out, that the program ran only on the headnode (not qdr-partition). 

With "salloc -p qdr -N 2 mpirun -n 2 hello_world_distributed --hpx:threads 2 --hpx:debug-clp" we get the following output:
[debug.txt](https://github.com/STEllAR-GROUP/hpx/files/12186385/debug.txt)

The command for showing the hostname ("salloc -p qdr -N 2 mpirun -n 2 hostname") only shows the headnode two times.

**Reply by pauleonix** [reply]
I now tried using `sbatch` and it seems to work without a problem, so this might be an issue with `salloc` specifically or rather with running `mpirun` from the headnode instead of from one of the compute nodes:
```console
$ cat test.sh
#!/bin/sh

spack load hpx

mpirun hostname

mpirun hello_world_distributed
$ sbatch -p qdr -N 2 test.sh
Submitted batch job 36856
$ # ... wait for completion ...
$ cat slurm-36856.out
ziti-qdr008.ziti.uni-heidelberg.de
ziti-qdr009.ziti.uni-heidelberg.de
hpx::init: command line warning: --hpx:localities used when running with SLURM, requesting a different number of localities (2) than have been assigned by SLURM (1), the application might not run properly.
hello world from OS-thread 0 on locality 0
hello world from OS-thread 0 on locality 1
```

**Reply by hkaiser** [reply]
BTW, you should be able to add `--hpx:ignore-batch-env` to the command line to avoid the warning.

**Reply by dkeyGit** [reply]
On this way it also works with our own compiled version of the hello_world_distribution program.

**Comment by hkaiser** [comment]
Try:
```
salloc -p qdr -N 2 mpirun -n 2 ./bin/hello_world_distriuted
```
This all depends on how your slurm/MPI integration was built and configured. We might have to try different things.
